src/two_stage_method.jl

```diff
@@ -78,24 +93,33 @@ function two_stage_method(prob::DiffEqBase.DEProblem,tpoints,data;kernel= :Epane
     f = prob.f
     n = length(tpoints)
     h = (n^(-1/5))*(n^(-3/35))*((log(n))^(-1/16))
-    estimated_solution = zeros(n,size(data)[1])
-    estimated_derivative = zeros(n,size(data)[1])
+    estimated_solution = zeros(size(data)[1],n)
+    estimated_derivative = zeros(size(data)[1],n)
     kernel_function = decide_kernel(kernel)
     e1 = [1;0]
     e2 = [0;1;0]
+    du = similar(prob.u0) # have to adjust type for autodifferentiation
+    sol = Vector{typeof(du)}(undef,n)
+    if mpg_autodiff
+      du_cache = DiffCache(du)
+      sol_cache = Vector{typeof(du_cache)}(undef,n)
+    end
     construct_estimated_solution_and_derivative!(estimated_solution,estimated_derivative,e1,e2,data,kernel_function,tpoints,h,n)
-
     # Step - 2
     cost_function = function (p)
-        # have to adjust type for autodifferentiation
-        du = similar(prob.u0, promote_type(eltype(prob.u0), eltype(p)))
-        sol = Vector{typeof(du)}(undef,n)
+        if mpg_autodiff
+          du_ = get_tmp(du_cache,eltype(p))
+          du = reinterpret(eltype(p),du_)
+          sol_ = map(i -> get_tmp(i,eltype(p)),sol_cache)
```

Comment:
@ChrisRackauckas  here I am getting `LoadError: UndefRefError: access to undefined reference` which I am not sure ow to avoid, is it because of using `undef` in the `Vector` initialisation?

Reply:
yes.